rasa/core/brokers/kafka.py

```diff
@@ -209,16 +218,65 @@ def _publish(self, event: Dict[Text, Any]) -> None:
             f" key={partition_key!s}, headers={headers})"
         )
 
+        serialized_event = json.dumps(event).encode(DEFAULT_ENCODING)
+
         if self.producer is not None:
-            self.producer.send(
-                self.topic, value=event, key=partition_key, headers=headers
+            self.producer.poll(0.0)
+            self.producer.produce(
+                self.topic,
+                value=serialized_event,
+                key=partition_key,
+                headers=headers,
+                on_delivery=delivery_report,
             )
 
     def _close(self) -> None:
-        if self.producer is not None:
-            self.producer.close()
+        self._cancelled = True
+        self._poll_thread.join()
 
     @rasa.shared.utils.common.lazy_property
     def rasa_environment(self) -> Optional[Text]:
         """Get value of the `RASA_ENVIRONMENT` environment variable."""
         return os.environ.get("RASA_ENVIRONMENT", "RASA_ENVIRONMENT_NOT_SET")
+
+    def _poll_loop(self) -> None:
+        if self.producer is not None:
+            while not self._cancelled:
+                self.producer.poll(0.1)
+
+
+def kafka_error_callback(err: "KafkaError") -> None:
+    """Callback for Kafka errors.
+
+    Any exception raised from this callback will be re-raised from the
+    triggering flush() call.
+    """
+    from confluent_kafka import KafkaException, KafkaError
+
+    # handle authentication / connection related issues, likely pointing
+    # to a configuration error
+    if (
+        err.code() == KafkaError._ALL_BROKERS_DOWN
+        or err.code() == KafkaError._AUTHENTICATION
+        or err.code() == KafkaError._MAX_POLL_EXCEEDED
+    ):
+        raise KafkaException(err)
+    else:
+        logger.error("A KafkaError has been raised.", exc_info=True)
```

Comment:
since we might be able to recover (according to the docs, that is what the confluent lib will try to do), i'd make this a warning. in `delivery_report` on the other hand i'd make the warning about the failed delivery an `error`, because we can't recover from it.

Reply:
Addressed: https://github.com/RasaHQ/rasa/pull/11765/commits/c34bc473b75fa51cffed3e1222c224633665f89e